XenModel/Actions/ParallelAction.cs

```diff
@@ -44,58 +44,128 @@ namespace XenAdmin.Actions
+            : this(title, startDescription, endDescription, subActions, false, false, maxNumberOfParallelActions)
+        { }
+
+        private void GroupActionsByConnection()
         {
-            foreach (AsyncAction subAction in subActions)
+            foreach (AsyncAction action in subActions)
             {
-                AsyncAction action = subAction;
-                action.Completed += action_Completed;
-                _queuePC.EnqueueItem(
-                    () =>
+                if (action.Connection != null)
+                {
+                    if (action.Connection.IsConnected)
                     {
-                        try
+                        if (!actionsByConnection.ContainsKey(action.Connection))
                         {
-                            action.RunExternal(Session);
+                            actionsByConnection.Add(action.Connection, new List<AsyncAction>());
                         }
-                        catch (Exception e)
-                        {
-                            Failure f = e as Failure;
-                            if (f != null && Connection != null &&
-                                f.ErrorDescription[0] == Failure.RBAC_PERMISSION_DENIED)
-                            {
-                                Failure.ParseRBACFailure(f, Connection, Session ?? Connection.Session);
-                            }
-                            exceptions.Add(e);
-                            // Record the first exception we come to. Though later if there are more than one we will replace this with non specific one.
-                            if (Exception == null)
-                                Exception = e;
-                        }
-                    });
 
+                        actionsByConnection[action.Connection].Add(action);
+                    }
+                }
+                else
+                {
+                    actionsWithNoConnection.Add(action);
+                }
+            }
+        }
+
+        private void CreateQueues(int maxNumberOfParallelActions)
+        {
+            foreach (IXenConnection connection in actionsByConnection.Keys)
+            {
+                queuesByConnection[connection] = new ProduceConsumerQueue(Math.Min(maxNumberOfParallelActions, actionsByConnection[connection].Count));
             }
+
+            if (actionsWithNoConnection.Count > 0)
+                queueWithNoConnection = new ProduceConsumerQueue(Math.Min(maxNumberOfParallelActions, actionsWithNoConnection.Count));
+        }
+
+        protected override void RunSubActions(List<Exception> exceptions)
+        {
+            foreach (IXenConnection connection in actionsByConnection.Keys)
+            {
+                queuesByConnection[connection] = new ProduceConsumerQueue(Math.Min(DEFAULT_MAX_NUMBER_OF_PARALLEL_ACTIONS, actionsByConnection[connection].Count));

```

Comment:
The queue size should be maxNumberOfParallelActions not DEFAULT_MAX_NUMBER_OF_PARALLEL_ACTIONS.